barretenberg/cpp/src/barretenberg/vm/avm/trace/finalization.hpp

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include "barretenberg/common/assert.hpp"
+#include <cstdint>
+#include <iostream>
+
+namespace bb::avm_trace {
+
+/**
+ * @brief Iterates over the main trace and an event trace and performs an action.
+ * @details This function iterates on the main trace and an event trace and applies:
+ * - `fmap` when the main trace clk matches the event trace clk.
+ * - `fall` for all rows that are EXECUTION trace rows.
+ */
+template <typename S, typename D, typename M, typename A>
+void iterate_with_actions(const S& src, D& main_trace, M&& fmap, A&& fall)
```

Comment:
Is the D template used to avoid import of AvmFullRow?

Reply:
I changed it to `std::vector<AvmFullrow<FF>>`. It was both to avoid the full row and the vector (anything that can be iterated and has size() would've worked). In any case, we can live with this and will generalize when the need comes.